### PR TITLE
modules: Add fff mocking framework

### LINKED ISSUE #38643: Add a mocking framework to go along ztest
## Origin

https://github.com/meekrosoft/fff

## Purpose

Allows mocking C functions.

## Mode of integration

Module: from my experience it would be easier to update the fork. The module name should simply be `fff`
 
## Description

Zephyr is light on tests and while zmake provides a test framework it's difficult to do anything less than an integration test right now. FFF is a simple function mocking framework that involves just a single header. Having this would allow for much faster tests with less dependencies.

# Dependencies

None

## Revision

`7e09f07e5b262b1cc826189dc5057379e40ce886`

## License

Please use an SPDX identifier (https://spdx.org/licenses/), such as
``BSD-3-Clause``

### PR DESCRIPTION
**Pull request by yperess**
Issue #38643

Introduce a more powerful and well established mocking framework
into Zephyr.

Signed-off-by: Yuval Peress <peress@google.com>

Closes #38643

### CODE REVIEW

File: modules/fff/CMakeLists.txt
Comment (yashi) [inline modules/fff/CMakeLists.txt:9]: fff is AFAICS a header only library.  Then, you don't need any thing to build.  You can remove `zephyr_library()`.

File: modules/fff/CMakeLists.txt
Comment (yashi) [inline modules/fff/CMakeLists.txt:11]: I'd just write it out in one line like:
```cmake
zephyr_include_directories(${ZEPHYR_CURRENT_MODULE_DIR})
```
But I'm don't know there is a standard we should follow or not.

File: west.yml
Comment (yashi) [inline west.yml:54]: I wouldn't do that.  Why don't you just point to your repo or wait til either
1. merged upstream
2. forked under zephyrproject-rtos

Just in case if you want to point to your repo:
```yaml
    - name: fff
      url: https://github.com/yperess/fff
      revision: master
      path: modules/lib/fff
```

File: modules/fff/Kconfig
Comment (yashi) [inline modules/fff/Kconfig:1]: I think it's better to put this file under `zephyr/` in your repo `yperess/fff` along with `module.yml`

File: modules/fff/CMakeLists.txt
Comment (yashi) [inline modules/fff/CMakeLists.txt:1]: I think it's better to put this file under `zephyr/` in your repo `yperess/fff` along with `module.yml` and `Kconfig` below.

**Review by yashi — CHANGES_REQUESTED** [review]
I'm not expert but I don't think you need anything under `zephyr/modules` if the upstream is willing to merge your changes.

**Comment by yashi** [comment]
Just for a reference, this is what I did locally.

````diff
From b40f45dfc68a1d1df062efcd0722ea8d11765118 Mon Sep 17 00:00:00 2001
From: Yasushi SHOJI <yashi@spacecubics.com>
Date: Thu, 23 Sep 2021 00:28:51 +0900
Subject: [PATCH] Add support for Zephyr Module

This commit is just to show you how I did it.

Signed-off-by: Yasushi SHOJI <yashi@spacecubics.com>
---
 zephyr/CMakeLists.txt | 7 +++++++
 zephyr/Kconfig        | 7 +++++++
 zephyr/module.yml     | 6 ++----
 3 files changed, 16 insertions(+), 4 deletions(-)
 create mode 100644 zephyr/CMakeLists.txt
 create mode 100644 zephyr/Kconfig

diff --git a/zephyr/CMakeLists.txt b/zephyr/CMakeLists.txt
new file mode 100644
index 0000000..745c81b
--- /dev/null
+++ b/zephyr/CMakeLists.txt
@@ -0,0 +1,7 @@
+# Copyright (c) 2021 Google LLC
+#
+# SPDX-License-Identifier: Apache-2.0
+
+if(CONFIG_FFF)
+  zephyr_include_directories(${ZEPHYR_CURRENT_MODULE_DIR})
+endif()
diff --git a/zephyr/Kconfig b/zephyr/Kconfig
new file mode 100644
index 0000000..6ed3db2
--- /dev/null
+++ b/zephyr/Kconfig
@@ -0,0 +1,7 @@
+# Copyright (c) 2021 Google LLC
+# SPDX-License-Identifier: Apache-2.0
+
+config FFF
+	bool "Fake Function Framework"
+	help
+	  This option enables the FFF mocking library.
diff --git a/zephyr/module.yml b/zephyr/module.yml
index 0507ce4..cbff6a1 100644
--- a/zephyr/module.yml
+++ b/zephyr/module.yml
@@ -1,5 +1,3 @@
-name: fff
 build:
-  cmake-ext: True
-  kconfig-ext: True
-
+  cmake: zephyr
+  kconfig: zephyr/Kconfig
-- 
2.33.0
````

With this installed on fff, All I needed on Zephyr is the following 4 lines in `west.yml`.
```
+    - name: fff
+      url: https://github.com/yperess/fff
+      revision: master
+      path: modules/lib/fff
```

File: modules/fff/CMakeLists.txt
Comment (yperess) [reply modules/fff/CMakeLists.txt:9]: You're right, done.

File: modules/fff/CMakeLists.txt
Comment (yperess) [reply modules/fff/CMakeLists.txt:11]: Done

File: west.yml
Comment (yperess) [reply west.yml:54]: https://docs.zephyrproject.org/latest/guides/modules.html#process-for-submitting-changes-to-existing-modules

File: modules/fff/Kconfig
Comment (yperess) [reply modules/fff/Kconfig:1]: You can see my reasoning here https://android-review.googlesource.com/c/platform/system/chre/+/1833040. It'll be the same as CHRE. I don't want the module maintainers to have to also worry about their downstream clients. We're point to a specific SHA and it is _our_ responsibility when we update the SHA to also make sure that things work.

File: modules/fff/CMakeLists.txt
Comment (yperess) [reply modules/fff/CMakeLists.txt:1]: Same as above.

File: west.yml
Comment (yashi) [reply west.yml:54]: Didn't know! Thanks.

File: tests/ztest/mock/src/main.c
Comment (yashi) [inline tests/ztest/mock/src/main.c:77]: a nitpick, don't need to change.

File: tests/ztest/mock/testcase.yaml
Comment (yashi) [inline tests/ztest/mock/testcase.yaml:3]: Care to tell us why?

**Review by yashi — COMMENTED** [review]
LGTM :+1:

File: tests/ztest/mock/src/main.c
Comment (yperess) [reply tests/ztest/mock/src/main.c:77]: Done

File: tests/ztest/mock/testcase.yaml
Comment (yperess) [reply tests/ztest/mock/testcase.yaml:3]: Yes, sorry, it looks like way back when ztest's mock was written it wasn't compatible with ARM. At the TSC meeting today the one primary concern was that FFF wasn't compatible with ARM. It seems to be building and running on the arm qemu so I wanted to enable the test.

Comment (yashi) [reply tests/ztest/mock/testcase.yaml:3]: Thanks.  Just for the note.  I run a few on my board with Cortex M3 as well.  So it seems working on ARM.

**Review by yashi — APPROVED** [review]

File: tests/ztest/mock/src/main.c
Comment (tejlmand) [inline tests/ztest/mock/src/main.c:83]: This appears messy to me.
Are you trying to test two different features (mock frameworks) inside the same test suite ?

Why not create a completely new test suite under: ` tests/ztest/mock_fff/`
or similar ?

**Review by tejlmand — CHANGES_REQUESTED** [review]

File: west.yml
Comment (yperess) [reply west.yml:25]: Why can't we use the upstream (assuming they accept the PR to include `zephyr/module.yml`)? We're fixing the SHA so even if they add a breaking change we would need to uprev it locally and test it so it's not going to break anything.

File: tests/ztest/mock/src/main.c
Comment (yperess) [reply tests/ztest/mock/src/main.c:83]: ACK, will split it into a separate suite, I wanted to put it in mock for discoverability.

File: tests/ztest/mock/src/test_suite_fff.h
Comment (yperess) [reply tests/ztest/mock/src/test_suite_fff.h]: Refactored into new test

File: tests/ztest/mock/src/main.c
Comment (yperess) [reply tests/ztest/mock/src/main.c]: Done, but I'm getting an error now for `native_posix` (will post as a separate comment)

**Comment by yperess** [comment]
I'm getting an error running twister for the `native_posix` platform:

```
/usr/bin/ld: ../app/libapp.a(main.c.obj): warning: relocation in read-only section `.text.test_main'
/usr/bin/ld: warning: creating DT_TEXTREL in a PIE
```

I can't quite figure it out, any help or pointer would be greatly appreciated.

**Comment by nashif** [comment]
material for thought: if all we need is this fff.h header, would it make sense to just include it in zephyr?

**Comment by yperess** [comment]
> material for thought: if all we need is this fff.h header, would it make sense to just include it in zephyr?

I would love to do that (much simpler) I just couldn't find Zephyr documentation on a way to uprev if an upstream change goes in. Do we just copy the header into `include` and override if we need to?

**Comment by yperess** [comment]
OK, I made quite a big change just now bringing in all the tests from FFF. They all pass on my local machine (waiting for CI) for every platform (very promising).

If this looks good to everyone else too, I'd like to update my PR to FFF to abstract away some of the testing constructs they have so we can remove `modules/fff/test/src` entirely and basically override the wrappers for the test framework.

Thoughts?

File: west.yml
Comment (tejlmand) [reply west.yml:25]: All modules must have a fork in Zephyr.
Please read the guide-lines:
https://docs.zephyrproject.org/latest/contribute/external.html#integration-as-a-module
https://docs.zephyrproject.org/latest/contribute/external.html#external-src-process
https://docs.zephyrproject.org/latest/guides/modules.html#process-for-submitting-a-new-module

**Comment by tejlmand** [comment]
> I'm getting an error running twister for the `native_posix` platform:
> 
> ```
> /usr/bin/ld: ../app/libapp.a(main.c.obj): warning: relocation in read-only section `.text.test_main'
> /usr/bin/ld: warning: creating DT_TEXTREL in a PIE
> ```
> 
> I can't quite figure it out, any help or pointer would be greatly appreciated.

Are you by any chance running `ld` : 2.35.2 ?
Debian 11 perhaps ?

File: west.yml
Comment (yperess) [reply west.yml:25]: I read those very carefully, the only mention of a fork is:

> If a module is maintained as a fork of another project on Github, the Zephyr module related files and changes in relation to upstream need to be maintained in a special branch named `zephyr`.

There's no hard requirement for it to be a fork neither from the documentation nor from a technical perspective. West is perfectly capable of handling more than one remote and instead of cluttering the `zephyrproject-rtos` with forks I would highly recommend we make use of this and first try to get upstream support in our modules. We can further discuss in this week's TSC meeting if you'd like.

Comment (tejlmand) [reply west.yml:25]: This has nothing to do with `west`.

It's an old decision that Zephyr project wants to be in control of all projects in the manifest.
Otherwise Zephyr bears the risk of a repo suddenly missing, a lost SHA (force push to upstream with history rewrite) or the project changing its license (in this case Zephyr can decide to stay at the version with an acceptable license).

Also note, for example a Zephyr LTS will be supported for long period, in which case there is an increased risk when depending on external repo.

With a fork, Zephyr community decides when to update, and in case something changes, whether or not to upmerge those changes.

Comment (yperess) [reply west.yml:25]: Those are valid concerns, can we update the documentation to reflect that? I wouldn't trust myself to get the wording right.

Comment (tejlmand) [reply west.yml:25]: > Those are valid concerns, can we update the documentation to reflect that? I wouldn't trust myself to get the wording right.

it should probably be documented (if it's not done so somewhere else).
But I don't trust my own wording on this either.

Comment (tejlmand) [reply west.yml:25]: it seems to be documented here:
https://docs.zephyrproject.org/latest/guides/modules.html#module-repositories

> Module Repositories
>
>   All modules included in the default manifest shall be hosted in repositories under the zephyrproject-rtos GitHub organization.

on the same page I linked earlier.

Comment (yperess) [reply west.yml:25]: I see, I was looking for the word `fork`, must have missed this. I'll keep working on the tests to make them cleaner and avoid having to copy them over (making good progress). Hopefully they'll be accepted upstream, if not then I'll just send them to the fork.

**Comment by yperess** [comment]
> Are you by any chance running `ld` : 2.35.2 ?
> Debian 11 perhaps ?

Yes, it seems that's is a Debian issue, if I comment out:
```
zephyr_cc_option(-fno-pie)
zephyr_cc_option(-fno-pic)
```

Then everything works. I guess I'll just need to do that locally when I run tests.

File: west.yml
Comment (yperess) [reply west.yml:25]: With the latest changes I'm very happy with this PR as it's able to run all of the C tests for FFF. @carlescufi what are your thoughts about creating the fork? We can merge my FFF PR there while it's being reviewed upstream and start using this. I want to make a big push for testing on Zephyr and this is a key aspect of it.

**Comment by nashif** [comment]
> > material for thought: if all we need is this fff.h header, would it make sense to just include it in zephyr?
> 
> I would love to do that (much simpler) I just couldn't find Zephyr documentation on a way to uprev if an upstream change goes in. Do we just copy the header into `include` and override if we need to?

yes, how often does this get updated? I would put it somewhere under subsys/testsuite/include in this case.. We probably should raise this as an option in the TSC and go fo it.

**Comment by yperess** [comment]
> > > material for thought: if all we need is this fff.h header, would it make sense to just include it in zephyr?
> > 
> > 
> > I would love to do that (much simpler) I just couldn't find Zephyr documentation on a way to uprev if an upstream change goes in. Do we just copy the header into `include` and override if we need to?
> 
> yes, how often does this get updated? I would put it somewhere under subsys/testsuite/include in this case.. We probably should raise this as an option in the TSC and go fo it.

Done, I've kept the tests in the module so we do need to get the fork done so I can update the `west.yml` file.

**Comment by yperess** [comment]
Looks like `fff.h` fails a bunch of the CI checks for style and guidelines. I'm open to any recommendations.

**Comment by yperess** [comment]
@nashif What would you recommend? It seem that bringing in `fff.h` directly will not work with the style guide. Should we:
1. Make an exception
2. Keep it in the module like before

**Comment by yperess** [comment]
Ping, any thoughts about how to move forward?

**Comment by jackrosenthal** [comment]
Can we re-format the source to match our style guides?

**Comment by yperess** [comment]
> Can we re-format the source to match our style guides?

I ran clang-format to get the formatting, but the CI complains about:
1. The license, since the file uses a different license (which I don't believe we're allowed to change).
2. The use of typedef, and a few lines that clang-format couldn't get under 100 lines.

I believe that the right approach here is to just make an exception and allow it as is. This file will not be updated often if at all so this will not be a recurring issue.

**Comment by jackrosenthal** [comment]
> > Can we re-format the source to match our style guides?
> 
> I ran clang-format to get the formatting, but the CI complains about:
> 
> The license, since the file uses a different license (which I don't believe we're allowed to change).

MIT is compatible with Apache 2. Adding an SPDX identifier should be sufficient.

> The use of typedef, and a few lines that clang-format couldn't get under 100 lines.

Can you please fix these? I don't think we want typedef'd structs.

**Review by tejlmand — CHANGES_REQUESTED** [review]
> For testing. It runs and builds everything with the Zephyr toolchain. This way of/when we update fff.h with upstream changes we can be sure it works still.

if we're anyway going to maintain our own fork that also contains the `fff.h` header then what's the value of bringing the same file into the Zephyr tree ?

I thought the reason to bring the file into Zephyr tree was to avoid creating a fork, as mentioned by @nashif here: https://github.com/zephyrproject-rtos/zephyr/pull/38730#issuecomment-927096456.
> material for thought: if all we need is this fff.h header, would it make sense to just include it in zephyr?

@nashif was the fact that we will now have the same file duplicated in two places:
- Zephyr repo
- fff repo fork

discussed at the TSC ?

I would prefer we only have such file in one place.

**Comment by olofj** [comment]
> > are we going to keep the external module? what for?
> 
> For testing. It runs and builds everything with the Zephyr toolchain. This way of/when we update `fff.h` with upstream changes we can be sure it works still.

This is not a good reason to include it for every Zephyr developer everywhere all the time.

The upstream repo seems to be very dormant, so a well-documented flow for how to test changes at time of import as part of the git history of fff.h seems to be a better approach here.

**Comment by yperess** [comment]
OK, I've added the `ci` group and filtered. What's the process now for:

1. Getting the fff fork made
2. Updating the CI to use the `ci` group so the module isn't excluded

**Comment by yperess** [comment]
> > * Updating the CI to use the ci group so the module isn't excluded
> 
> Why does CI need the module ? Isn't CI using the file that is now included with Zephyr here: https://github.com/zephyrproject-rtos/zephyr/blob/4cea3934a64451dc893e3d0db00c14597ffdb492/subsys/testsuite/include/fff.h

The CI runs the new test suite which is just a wrapper for the full test suite found on the fff module. This means we're actually running the full test suite of fff but using the Zephyr toolchain and ztest. The idea is that if we ever need to uprev `fff.h` we'll want to verify that nothing broke. The CI will be able to run all the tests. This is a much better alternative to copying the fff tests into Zephyr since we can now get upstream tests by simply syncing the fork.

**Comment by ElectricalPaul** [comment]
> > * Updating the CI to use the ci group so the module isn't excluded
> 
> Why does CI need the module ? Isn't CI using the file that is now included with Zephyr here: https://github.com/zephyrproject-rtos/zephyr/blob/4cea3934a64451dc893e3d0db00c14597ffdb492/subsys/testsuite/include/fff.h

fff.h is in the Zephyr tree, but the full FFF tests are not. The tests in this PR are just a wrapper (using Twister and the Zephyr toolchain) for the FFF tests.

We have a fork of FFF that contains all of the test suites. If FFF pushes new tests, we can update the SHA that our fork points to, and then we get all of their new tests.

**Review by ElectricalPaul — APPROVED** [review]

**Review by tejlmand — CHANGES_REQUESTED** [review]
It should always be possible to run twister locally on a clean Zephyr workspace without enabling extra manifest groups.

**Comment by nashif** [comment]
Trying to run this locally I get:

```
[  8%] Building C object CMakeFiles/app.dir/src/main.c.obj
/home/nashif/Work/zephyrproject/zephyr/tests/ztest/mock_fff/src/main.c:9:10: fatal error: c_test_framework.h: No such file or directory
    9 | #include <c_test_framework.h>
      |          ^~~~~~~~~~~~~~~~~~~~
compilation terminated.
gmake[2]: *** [CMakeFiles/app.dir/build.make:76: CMakeFiles/app.dir/src/main.c.obj] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:2055: CMakeFiles/app.dir/all] Error 2
```

not sure how it is passing CI.

**Comment by tejlmand** [comment]
> this was discussed in the TSC, so removing TSC label

@nashif it was re-added by @olofj here to be taken up again:
https://github.com/zephyrproject-rtos/zephyr/pull/38730#issuecomment-942351260

**Review by tejlmand — APPROVED** [review]
LGTM, @yperess thanks for the effort.

**Review by nashif — APPROVED** [review]

**Pull request merged**
